File: backend/app/schemas/financial.py
from pydantic import BaseModel, Field
from typing import List, Dict, Optional, Any

class PersonalDetails(BaseModel):
    age: int = Field(..., ge=18, le=100, example=32)
    gender: str = Field(..., example="Male")  # Male, Female, Other
    occupation: str = Field(..., example="Software Engineer")
    employment_type: str = Field(..., example="Salaried") # Salaried, Self-Employed, Business, Unemployed
    marital_status: str = Field(..., example="Married") # Single, Married, Divorced
    number_of_dependents: int = Field(..., ge=0, le=10, example=2)

class IncomeDetails(BaseModel):
    monthly_salary: float = Field(..., ge=0, example=85000.0)
    additional_income: float = Field(0.0, ge=0, example=5000.0)
    annual_bonus: float = Field(0.0, ge=0, example=120000.0)
    business_income: float = Field(0.0, ge=0, example=0.0)

class ExpenseDetails(BaseModel):
    house_rent: float = Field(0.0, ge=0, example=20000.0)
    food_expenses: float = Field(0.0, ge=0, example=12000.0)
    transportation: float = Field(0.0, ge=0, example=4000.0)
    electricity_bill: float = Field(0.0, ge=0, example=2500.0)
    internet_bill: float = Field(0.0, ge=0, example=1000.0)
    insurance: float = Field(0.0, ge=0, example=3000.0)
    education: float = Field(0.0, ge=0, example=5000.0)
    medical_expenses: float = Field(0.0, ge=0, example=2000.0)
    entertainment: float = Field(0.0, ge=0, example=3500.0)
    shopping: float = Field(0.0, ge=0, example=6000.0)
    other_expenses: float = Field(0.0, ge=0, example=3000.0)

class LoanDetails(BaseModel):
    loan_type: str = Field("Home Loan", example="Personal Loan")
    loan_amount: float = Field(0.0, ge=0, example=500000.0)
    interest_rate: float = Field(10.5, ge=0, le=50, example=10.5)
    loan_tenure_months: int = Field(36, ge=6, le=360, example=36)
    existing_emi: float = Field(0.0, ge=0, example=5000.0)
    credit_card_emi: float = Field(0.0, ge=0, example=2000.0)
    personal_loan_emi: float = Field(0.0, ge=0, example=0.0)
    home_loan_emi: float = Field(0.0, ge=0, example=15000.0)
    vehicle_loan_emi: float = Field(0.0, ge=0, example=0.0)

class BankingDetails(BaseModel):
    credit_score: int = Field(750, ge=300, le=900, example=750)
    bank_balance: float = Field(150000.0, ge=0, example=150000.0)
    savings: float = Field(200000.0, ge=0, example=200000.0)
    investments: float = Field(300000.0, ge=0, example=300000.0)

class LifestyleDetails(BaseModel):
    smoking: str = Field("No", example="No") # Yes, No
    alcohol: str = Field("Occasionally", example="Occasionally") # Never, Frequently, Linearly, Occasionally
    travel_frequency: str = Field("Moderate", example="Moderate") # Low, Moderate, High
    online_shopping_frequency: str = Field("High", example="High") # Low, Moderate, High

class FinancialProfileInput(BaseModel):
    personal: PersonalDetails
    income: IncomeDetails
    expenses: ExpenseDetails
    loans: LoanDetails
    banking: BankingDetails
    lifestyle: LifestyleDetails

class FeatureImportanceItem(BaseModel):
    feature: str
    impact: float
    description: str
    direction: str  # positive, negative, neutral

class RecommendationItem(BaseModel):
    category: str
    severity: str # High, Medium, Low, Positive
    title: str
    description: str
    action_item: str

class FutureTrendItem(BaseModel):
    month: str
    predicted_spending: float
    predicted_savings: float
    confidence_upper: float
    confidence_lower: float

class PredictionResponse(BaseModel):
    predicted_monthly_emi: float
    recommended_max_emi: float
    monthly_savings_after_emi: float
    disposable_income: float
    debt_to_income_ratio: float
    financial_health_score: float # 0 to 100
    loan_approval_chance: float # 0 to 100%
    emi_default_risk: float # 0 to 100%
    budget_score: float # 0 to 100
    investment_capacity: float
    future_trends: List[FutureTrendItem]
    top_feature_impacts: List[FeatureImportanceItem]
    recommendations: List[RecommendationItem]

File: frontend/src/types/financial.ts
export interface PersonalDetails {
  age: number;
  gender: string;
  occupation: string;
  employment_type: string;
  marital_status: string;
  number_of_dependents: number;
}

export interface IncomeDetails {
  monthly_salary: number;
  additional_income: number;
  annual_bonus: number;
  business_income: number;
}

export interface ExpenseDetails {
  house_rent: number;
  food_expenses: number;
  transportation: number;
  electricity_bill: number;
  internet_bill: number;
  insurance: number;
  education: number;
  medical_expenses: number;
  entertainment: number;
  shopping: number;
  other_expenses: number;
}

export interface LoanDetails {
  loan_type: string;
  loan_amount: number;
  interest_rate: number;
  loan_tenure_months: number;
  existing_emi: number;
  credit_card_emi: number;
  personal_loan_emi: number;
  home_loan_emi: number;
  vehicle_loan_emi: number;
}

export interface BankingDetails {
  credit_score: number;
  bank_balance: number;
  savings: number;
  investments: number;
}

export interface LifestyleDetails {
  smoking: string;
  alcohol: string;
  travel_frequency: string;
  online_shopping_frequency: string;
}

export interface FinancialProfileInput {
  personal: PersonalDetails;
  income: IncomeDetails;
  expenses: ExpenseDetails;
  loans: LoanDetails;
  banking: BankingDetails;
  lifestyle: LifestyleDetails;
}

export interface FeatureImportanceItem {
  feature: string;
  impact: number;
  description: string;
  direction: 'positive' | 'negative' | 'neutral';
}

export interface RecommendationItem {
  category: string;
  severity: 'High' | 'Medium' | 'Low' | 'Positive';
  title: string;
  description: string;
  action_item: string;
}

export interface FutureTrendItem {
  month: string;
  predicted_spending: number;
  predicted_savings: number;
  confidence_upper: number;
  confidence_lower: number;
}

export interface PredictionResponse {
  predicted_monthly_emi: number;
  recommended_max_emi: number;
  monthly_savings_after_emi: number;
  disposable_income: number;
  debt_to_income_ratio: number;
  financial_health_score: number;
  loan_approval_chance: number;
  emi_default_risk: number;
  budget_score: number;
  investment_capacity: number;
  future_trends: FutureTrendItem[];
  top_feature_impacts: FeatureImportanceItem[];
  recommendations: RecommendationItem[];
}

export interface DashboardAnalytics {
  benchmark_expense_breakdown: { category: string; percentage: number; amount: number }[];
  credit_score_risk_curve: { credit_score: number; default_risk: number; approval_rate: number }[];
  tenure_emi_comparison: { tenure_months: number; monthly_emi: number; total_interest: number }[];
}

export interface ModelMetricsSummary {
  best_regressor: string;
  best_classifier: string;
  regression_metrics: Record<string, { cv_r2_mean: number; mae: number; rmse: number; r2_score: number; mape: number }>;
  classification_metrics: Record<string, { cv_f1_mean: number; accuracy: number; precision: number; recall: number; f1_score: number; roc_auc: number }>;
  total_samples: number;
  features_count: number;
}
